File: src/main/java/epicsquid/superiorshields/setup/compat/MalumCompat.java
package epicsquid.superiorshields.setup.compat;

import com.tterrag.registrate.util.nullness.NonNullFunction;
import epicsquid.superiorshields.item.SoulStainedSteelShield;
import epicsquid.superiorshields.item.SpiritHunterShield;
import epicsquid.superiorshields.item.SuperiorShieldItem;
import epicsquid.superiorshields.item.VanillaShieldItem;
import epicsquid.superiorshields.shield.IShieldType;
import net.minecraft.world.item.Item;
import net.minecraftforge.fml.ModList;

public class MalumCompat {

	private static boolean LOADED;

	public static void init() {
		LOADED = ModList.get().isLoaded("malum");
	}

	public static NonNullFunction<Item.Properties, SuperiorShieldItem<IShieldType>> makeSoulStainedShieldOrDefault(IShieldType type) {
		return LOADED ? LoadedOnly.makeSoulStainedShield(type) : props -> new VanillaShieldItem(props, type);
	}

	public static NonNullFunction<Item.Properties, SuperiorShieldItem<IShieldType>> makeSpiritHunterShieldOrDefault(IShieldType type) {
		return LOADED ? LoadedOnly.makeSpiritHunterShield(type) : props -> new VanillaShieldItem(props, type);
	}


	public static class LoadedOnly {

		public static NonNullFunction<Item.Properties, SuperiorShieldItem<IShieldType>> makeSoulStainedShield(IShieldType type) {
			return props -> new SoulStainedSteelShield(props, type);
		}

		public static NonNullFunction<Item.Properties, SuperiorShieldItem<IShieldType>> makeSpiritHunterShield(IShieldType type) {
			return props -> new SpiritHunterShield(props, type);
		}
	}
}

File: src/main/kotlin/dev/epicsquid/superiorshields/compat/MalumCompat.kt
package dev.epicsquid.superiorshields.compat

import dev.epicsquid.superiorshields.item.SoulStainedSuperiorShieldItem
import dev.epicsquid.superiorshields.item.SpiritHunterSuperiorShieldItem
import dev.epicsquid.superiorshields.item.SuperiorShieldItem
import dev.epicsquid.superiorshields.shield.DurabilitySuperiorShield
import net.minecraft.world.item.Item.Properties
import net.minecraft.world.item.crafting.Ingredient
import net.minecraftforge.fml.ModList

object MalumCompat {

	private val loaded: Boolean by lazy { ModList.get().isLoaded("malum") }

	fun spiritHunterShieldBuilder(
		props: Properties,
		enchantmentValue: Int,
		type: DurabilitySuperiorShield,
		repairItem: () -> Ingredient?,
	): SuperiorShieldItem<DurabilitySuperiorShield> =
		if (loaded) {
			LoadedOnly.spiritHunterShieldBuilder(props, enchantmentValue, type, repairItem)
		} else {
			SuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = type,
				repairItem = repairItem
			)
		}

	fun soulStainedShieldBuilder(
		props: Properties,
		enchantmentValue: Int,
		type: DurabilitySuperiorShield,
		repairItem: () -> Ingredient?,
	): SuperiorShieldItem<DurabilitySuperiorShield> =
		if (loaded) {
			LoadedOnly.soulStainedShieldBuilder(props, enchantmentValue, type, repairItem)
		} else {
			SuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = type,
				repairItem = repairItem
			)
		}

	object LoadedOnly {
		fun spiritHunterShieldBuilder(
			props: Properties,
			enchantmentValue: Int,
			type: DurabilitySuperiorShield,
			repairItem: () -> Ingredient?,
		): SuperiorShieldItem<DurabilitySuperiorShield> =
			SpiritHunterSuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = type,
				repairItem = repairItem,
			)

		fun soulStainedShieldBuilder(
			props: Properties,
			enchantmentValue: Int,
			type: DurabilitySuperiorShield,
			repairItem: () -> Ingredient?,
		): SuperiorShieldItem<DurabilitySuperiorShield> =
			SoulStainedSuperiorShieldItem(
				props = props,
				enchantmentValue = enchantmentValue,
				type = type,
				repairItem = repairItem
			)
	}
}